docs/designs/keras_callback.md

```diff
@@ -0,0 +1,196 @@
+
+1. Only the parameter server calls `on_train_batch_begin|end` and the worker doesn't. Because, we 
+have not observed a use case for callback to execute per batch by the worker.
+
+2. ElasticDL sets an argmuent `instance_role` in `params` of each callback and users get the `instance_role` in `on_train_batch_begin|end`. It may need users to understand why to get the `instance_role`.
+```python
+class LearningRateScheduler(tf.keras.callbacks.Callback):
+    def on_train_batch_begin(self, batch, logs=None):
+        instance_role = self.params.get("instance_role", None)
+        if instance_role != "ParameterServer":
+            return
+```
+
+
+### Call Methods of Callback Per Epoch
+
+The methods should be called per epoch in `tf.keras.callbacks.Callback` is
+```python
+on_epoch_begin(self, epoch, logs=None)
+on_epoch_end(self, epoch, logs=None)
+```
+Only the master knows when an epoch begins and ends because the master creates tasks for each epoch using `_TaskDispatcher`. So, we can call those methods in the `get` and `report` of `_TaskDispatcher`.
+```python
+def get(self, worker_id):
+    if not self._todo and self._epoch < self._num_epochs - 1:
+        # Start a new epoch
+        self._epoch += 1
+        self.create_tasks(elasticdl_pb2.TRAINING)
+        logger.info("Starting epoch %d", self._epoch)
+        self.callback_list.on_epoch_begin(self._epoch, logs)
+        
+        task = self._todo.pop()
+        if not self._todo:
+            self._epoch_end_task = task
+
+
+def report(self, request, success):
+    task_id = request.task_id
+    if task_id == self._epoch_end_task and success:
+        self.callback_list.on_epoch_end(self._epoch, logs)
+```
+
+
+### Call Methods of Callback When the Job Begins and Ends
+
+The methods should be called when the job begins and ends in `tf.keras.callbacks.Callback`
+```python
+on_(train|test|predict)_begin(self, logs=None)
+on_(train|test|predict)_end(self, logs=None)
+```
+The master starts and finishes a job, so it calls those methods. 
+
+```python
+def run(self):
+    """
+    The main loop of master.
+    Dispatch the tasks to the workers until all the tasks are completed.
+    """
+    try:
+        self.callback_list.on_(train|test|predict)_begin(self, logs=None)
+        while True: 
+            ...
+    except KeyboardInterrupt:
+        self.logger.warning("Server stopping")
+    finally:
+        self.callback_list.on_(train|test|predict)_end(self, logs=None)
+        self._stop()
+    return self._exit_code
+```
+
+
+### Call the Methods of Callback using AllReduceStrategy
+
+The design to call the methods using AllReduceStrategy will be considered later.
```

Comment:
This case would be even simpler - there are only master and workers.

Reply:
Yeah, the worker can call all methods in `tf.keras.callbacks.Callback` using AllReduceStrategy.